multimodal/vl2l/src/mlperf_inference_multimodal_vl2l/evaluation.py

```diff
@@ -0,0 +1,184 @@
+
+        total_intersection += intersection
+        total_predicted_length += pred_len
+        total_true_length += true_len
+
+    # 2. Calculate hP and hR
+    hp = total_intersection / total_predicted_length \
+        if total_predicted_length > 0 else 0.0
+    hr = total_intersection / total_true_length \
+        if total_true_length > 0 else 0.0
+
+    return 0.0 if hp + hr == 0 else 2 * (hp * hr) / (hp + hr)
+
+
+def calculate_exact_match(generated_text: str, original_text: str) -> float:
+    """Calculates binary Exact Match (EM) score.
+
+    We clean the text (lowercase, strip whitespace) for a fairer comparison.
+
+    Args:
+        generated_text: Output from the VLM.
+        original_text: Ground truth information from the dataset.
+
+    Returns:
+        1 if the values match or 0 otherwise
+    """
+    gen = generated_text.strip().lower()
+    orig = original_text.strip().lower()
+
+    return 1.0 if gen == orig else 0.0
+
+
+def alt_f1_score(data: list[tuple[str, str]]) -> float:
+    """Alt method to calculate hierarchical F1."""
+    y_pred_raw = []
+    y_true_raw = []
+
+    for pred, src in data:
+        path1 = pred.split(" > ")
+        path2 = src.split(" > ")
+
+        y_pred_raw.append(path1)
+        y_true_raw.append(path2)
+
+    # 2. Find the global maximum length across ALL samples
+    # We check the longest path in both true and pred lists
+    max_len = max(len(p) for p in y_true_raw + y_pred_raw)
+
+    # 3. Pad all lists to the global max_len
+    for i in range(len(y_true_raw)):
+        # Pad Truth
+        pad_len_true = max_len - len(y_true_raw[i])
+        y_true_raw[i] += [""] * pad_len_true
+
+        # Pad Prediction
+        pad_len_pred = max_len - len(y_pred_raw[i])
+        y_pred_raw[i] += [""] * pad_len_pred
+
+    # 4. Convert to numpy arrays
+    y_true = np.array(y_true_raw)
+    y_pred = np.array(y_pred_raw)
+
+    # 5. Calculate Score
+    return f1(y_true, y_pred)
+
+
+def run_evaluation(filename: FilePath, dataset: "DatasetCLI") -> None:
+    """Main function to run the evaluation."""
+    with Path.open(filename) as f:
+        model_output = json.load(f)
+
+    original_data = load_dataset(
+        dataset.repo_id,
+        dataset.token,
+    )[dataset.split]
+
+    category_dataset_pred_src = []
+    is_secondhand_pred = []
+    is_secondhand_src = []
+    for elem in model_output:
+        byte_data = bytes.fromhex(elem["data"])
+        idx = elem["qsl_idx"]
+        pred_text_decode = byte_data.decode("utf-8")
+        pred_item = json.loads(pred_text_decode)
+        ground_truth_item = original_data[idx]
+        category_dataset_pred_src.append((pred_item["category"],
+                                          ground_truth_item["ground_truth_category"]))
+        is_secondhand_pred.append(int(pred_item["is_secondhand"]))
+        is_secondhand_src.append(
+            int(ground_truth_item["ground_truth_is_secondhand"]))
```

Comment:
I don't have a preference between having the inputs to the scoring function as a `list[tuple[str, str]]` or `list[str], list[str]`, but as a good practice, we should keep it consistent to be one or the other.